---
title: "Example TARA analyses"
author: "daniel.lundin@lnu.se"
date: "`r format(Sys.time(), '%Y-%m-%d')`"
output:
  html_document:
    toc: yes
    toc_float:
      collapse: no
    fig_caption: yes
    code_folding: hide
  pdf_document:
    fig_caption: yes
    fig_height: 9
    fig_width: 8
    number_sections: yes
    toc: yes
bibliography: references.bib
---

```{r setup, echo=F, cache = FALSE}
knitr::opts_chunk$set(echo=F, fig.width = 12, fig.path='figures/', cache = TRUE)
ggplot2::theme_set(ggplot2::theme_bw())
```

```{r libraries, message=F, cache = FALSE}
suppressPackageStartupMessages(library(readr))
suppressPackageStartupMessages(library(feather))
suppressPackageStartupMessages(library(dplyr, warn.conflicts = FALSE))
suppressPackageStartupMessages(library(tidyr))
suppressPackageStartupMessages(library(purrr))
suppressPackageStartupMessages(library(stringr))
suppressPackageStartupMessages(library(ggplot2))
suppressPackageStartupMessages(library(ggrepel))
suppressPackageStartupMessages(library(kfigr))
suppressPackageStartupMessages(library(knitr))
```

```{r constants}
```

```{r read-idxstats-data}
# Read all idxstats feather files
idxstats <- tibble(fn = Sys.glob("data/idxstats.*.feather")) %>% # Create a table with all file names
  mutate(
    d = map(fn, read_feather) # Read each idxstats table into a *cell*
  ) %>% 
  unnest(d) %>% # Expand each table from its cell
  mutate(sample = str_remove(sample, "\\.genomes.*")) %>% # Remove trailing characters from sample name
  select(-fn) # Get rid of the file name column
```

```{r read-samplecounts}
samplecounts <- tibble(fn = Sys.glob("data/samplecounts.*.tsv")) %>%
  mutate(
    d = map(fn, function(f) read_tsv(f, col_names = c('sample', 'size'), col_types = 'ci'))
  ) %>%
  unnest(d) %>%
  select(-fn) %>%
  # Get rid of duplicates. (They shouldn't be in the input.)
  distinct()
```

```{r calc-relab}
# To calculate relative abundances, i.e. counts from idxstats divided by total library sizes,
# join the two tables and create a new column.
# Assign the result to a new table with the same name as the old, overwriting the old.
idxstats <- idxstats %>%
  inner_join(samplecounts, by = 'sample') %>%
  mutate(relab = n_mapped/size) %>%
  select(-size)
```

```{r read-ref-data}
# Read the samples table
samples <- read_feather("data/simple_sample_table.feather")

# Read the genome index
gindex <- read_tsv(
  "data/genomeindex.tsv",
  col_names = c("contig", "genome"),
  col_types = "cc"
)
```

```{r check-samples-table, include = FALSE}
# Find samples in the idxstats table that are not present in the samples table
idxstats %>% distinct(sample) %>%
  anti_join(samples, by = "sample")
```

```{r create-calothrix-table}
# Start with the genome table, subset to genomes matching "calothrix"
calothrix <- gindex %>% filter(grepl('Calothrix', genome)) %>%
  # Join in the table with relative abundances
  inner_join(idxstats, by = c('contig' = 'sequence')) %>%
  # Sum relative abundances by genome and sample, to get rid of the contig data
  group_by(genome, sample) %>%
  summarise(relab = sum(relab), .groups = 'drop') %>%
  # Join in the sample table to get the latitude and depth_code.
  inner_join(
    # Filter the samples table to only contain the top three depths
    samples %>% filter(depth_code %in% c('SRF', 'DCM', 'MES')), 
    by = 'sample'
  )
```

```{r create-topsamples}
topsamples <- gindex %>% filter(genome == 'Calothrix rhizosoleniae SC01' | grepl('Richelia intra', genome)) %>%
  inner_join(idxstats, by = c('contig' = 'sequence')) %>%
  group_by(genome, sample) %>% summarise(relab = sum(relab), .groups = 'drop_last') %>%
  top_n(n = 5, wt = relab) %>%
  ungroup() %>%
  distinct(sample)
```

# Introduction

```{r station-map, fig.width = 12, fig.height = 8, fig.cap = '**TARA stations.**', warning = FALSE}
samples %>% group_by(station) %>% summarise(lat = min(lat), long = min(long)) %>%
  ggplot(aes(x = long, y = lat)) +
  geom_point() +
  borders('world') +
  geom_text_repel(aes(label = station), size = 2, max.overlaps = 10) 
```


TEXT [@gloor2017]

[A blog post about citations](https://blog.rstudio.com/2020/11/09/rstudio-1-4-preview-citations/)

```{r}
# I think ... above
```

# Materials and methods

Something about how we did the mapping of genomic TARA reads to genomes.

# Results

Something about the below plot (`r figr('genome-overview', T, type = 'Figure')`).

```{r genome-overview, fig.width = 12, fig.cap='**Total relative abundance per genome.** *Note*, numbers were not divided by the number of samples; i.e. not means.'}
# Create a plot of how much each genome attracted in total in the different depth classes.
idxstats %>%
  # Join in samples to get the depth code for each sample
  inner_join(samples, by = "sample") %>%
  # Join in gindex to get the name of the organism
  inner_join(gindex, by = c("sequence" = "contig")) %>%
  # To sum, group by the variables to sum over...
  group_by(genome, depth_code) %>% 
  # then do the sum. (The .groups argument avoids a warning and leaves the data in a sortable, useful state.)
  summarise(relab = sum(relab), .groups = 'drop') %>%
  # Pass the data to the plotting functions; specify which variables you want to map to what part of the plot.
  ggplot(aes(x = genome, y = relab, colour = depth_code)) +
  # Do a dot plot
  geom_point() +
  # Use a log scale so that both large and small numbers can be seen.
  scale_y_log10() +
  xlab('Genome') + ylab('Relative abundance') +
  # Flip the coordinate system to make genome labels easier to read.
  coord_flip()
```

```{r richelia-latitude-dependency, fig.height = 8, fig.cap = "**Abundance of *Richelia* as a function of latitude.**"}
# Start with the genome table, subset to genomes matching "Richelia"
gindex %>% filter(grepl('Richelia', genome)) %>%
  # Join in the table with relative abundances
  inner_join(idxstats, by = c('contig' = 'sequence')) %>%
  # Sum relative abundances by genome and sample, to get rid of the contig data
  group_by(genome, sample) %>%
  summarise(relab = sum(relab), .groups = 'drop') %>%
  # Join in the sample table to get the latitude and depth_code.
  inner_join(
    # Filter the samples table to only contain the top three depths
    samples %>% filter(depth_code %in% c('SRF', 'DCM', 'MES')), 
    by = 'sample'
  ) %>%
  # We have all information, pass to ggplot and decide which type of data to plot where
  ggplot(aes(x = abs(lat), y = relab, colour = depth_code)) +
  # Make it a point plot
  geom_point() +
  # Create one subplot for each combination of genome and depth_code
  facet_wrap(depth_code~genome, ncol = 8) +
  xlab('abs(latitude)') + ylab('Relative abundance') +
  # Move the colour legend to the bottom of the plot to give space for a longer x axis
  theme(legend.position = 'bottom')
```

```{r calothrix-latitude-dependency, fig.height = 8, fig.cap = "**Abundance of *Calothrix* as a function of latitude.**"}
calothrix %>%
  # We have all information, pass to ggplot and decide which type of data to plot where
  ggplot(aes(x = abs(lat), y = relab, colour = depth_code)) +
  # Make it a point plot
  geom_point() +
  # Create one subplot for each combination of genome and depth_code
  facet_wrap(depth_code~genome, ncol = 8) +
  xlab('abs(latitude)') + ylab('Relative abundance') +
  # Move the colour legend to the bottom of the plot to give space for a longer x axis
  theme(legend.position = 'bottom')
```

```{r srf-dcm-correlation, fig.cap = "**Correlation between SRF and DCM.**"}
calothrix %>% 
  group_by(genome, station, lat = abs(round(lat)), depth_code) %>%
  summarise(mean_relab = mean(relab), .groups = 'drop') %>%
  pivot_wider(names_from = depth_code, values_from = mean_relab, values_fill = 0) %>%
  ggplot(aes(x = SRF, y = DCM, colour = lat)) +
  geom_point() +
  scale_colour_viridis_c()
```

```{r srf-mes-correlation, fig.width = 8, fig.height = 6, fig.cap = "**Correlation between SRF and DCM.**"}
calothrix %>% 
  group_by(genome, station, lat = abs(round(lat)), depth_code) %>%
  summarise(mean_relab = mean(relab), .groups = 'drop') %>%
  pivot_wider(names_from = depth_code, values_from = mean_relab, values_fill = 0) %>%
  ggplot(aes(x = SRF, y = MES, colour = lat)) +
  geom_point() +
  scale_colour_viridis_c()
```

```{r topsamples-map, fig.height = 12, fig.cap = "**Proportions of *Calothrix* and *Richelia* spp. in the samples where they were most abundant.**"}
gindex %>% 
  filter(genome == 'Calothrix rhizosoleniae SC01' | grepl('Richelia intra', genome)) %>%
  inner_join(idxstats, by = c('contig' = 'sequence')) %>%
  group_by(sample, genome) %>% summarise(relab = sum(relab), .groups = 'drop') %>%
  inner_join(samples, by = 'sample') %>%
  semi_join(topsamples, by = 'sample') %>%
  ggplot(aes(long, lat, size = relab)) +
  geom_point() +
  borders('world') +
  facet_grid(genome~depth_code)
```

```{r top-station-plot, fig.width=12, fig.height=5, fig.cap='**Top 15 stations for three Richelia spp. and Calothrix.**'}
gindex %>% 
  filter(genome %in% c('Calothrix rhizosoleniae SC01', 'Richelia intracellularis HH01', 'Richelia intracellularis HM01', 'Richelia intracellularis RC01')) %>%
  inner_join(idxstats, by = c('contig' = 'sequence')) %>%
  group_by(genome, sample) %>% summarise(relab = sum(relab) , .groups = 'drop_last') %>%
  top_n(15, wt = relab) %>%
  #pivot_wider(names_from = genome, values_from = relab) %>%
  inner_join(samples, by = 'sample') %>%
  ggplot(aes(long, lat, size = relab, colour = genome)) +
  geom_point(alpha = 0.5) +
  borders('world') +
  facet_wrap(~depth_code) +
  theme(legend.position = 'bottom')

```

# References
